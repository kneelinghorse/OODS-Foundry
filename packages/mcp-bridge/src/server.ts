import cors from '@fastify/cors';
import rateLimit from '@fastify/rate-limit';
import fastifyStatic from '@fastify/static';
import Fastify, { type FastifyReply, type FastifyRequest } from 'fastify';
import { spawn, ChildProcessWithoutNullStreams } from 'node:child_process';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { bridgeConfig, lowerCaseHeaders } from './config.js';
import { registerArtifactEndpoints } from './endpoints/artifacts.js';
import { buildErrorPayload, normalizeRunErrorCode, sendError, statusForCode } from './middleware/errors.js';

const ALLOWED_TOOLS = new Set(bridgeConfig.tools.allowed);
const WRITE_GATED_TOOLS = new Set(bridgeConfig.tools.writeGated);

type RunRequestBody = {
  tool?: string;
  input?: Record<string, any>;
};

type McpResponse =
  | { id?: string | number; result: any }
  | { id?: string | number; error: { code: string; message?: string; messages?: any } };

class McpClient {
  private child: ChildProcessWithoutNullStreams | null = null;
  private seq = 0;
  private pending = new Map<
    number,
    { resolve: (v: any) => void; reject: (e: any) => void }
  >();
  private buffer = '';

  constructor(private serverCwd: string) {}

  private ensure() {
    if (this.child && !this.child.killed) return;
    // Spawn the built MCP server (dist/index.js)
    const entry = path.join(this.serverCwd, 'dist', 'index.js');
    this.child = spawn(process.execPath, [entry], {
      cwd: this.serverCwd,
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env },
    });

    this.child.stdout.setEncoding('utf8');
    this.child.stdout.on('data', (chunk: string) => {
      this.buffer += chunk;
      let idx: number;
      while ((idx = this.buffer.indexOf('\n')) >= 0) {
        const line = this.buffer.slice(0, idx).trim();
        this.buffer = this.buffer.slice(idx + 1);
        if (!line) continue;
        try {
          const msg = JSON.parse(line) as McpResponse;
          const id = (msg as any).id as number | undefined;
          if (id != null && this.pending.has(id)) {
            const p = this.pending.get(id)!;
            this.pending.delete(id);
            if ((msg as any).error) p.reject((msg as any).error);
            else p.resolve((msg as any).result);
          }
        } catch {
          // ignore parse errors on stdout
        }
      }
    });

    this.child.on('exit', () => {
      // Reject all in-flight requests
      for (const [, p] of this.pending) p.reject({ code: 'PROCESS_EXIT' });
      this.pending.clear();
      this.child = null;
    });
  }

  async run(tool: string, input: any): Promise<any> {
    this.ensure();
    const id = ++this.seq;
    const payload = JSON.stringify({ id, tool, input }) + '\n';
    return new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      this.child!.stdin.write(payload, 'utf8');
    });
  }
}


function isJsonContentType(header: unknown): boolean {
  if (typeof header !== 'string') return false;
  const [mediaType] = header.split(';', 1);
  return mediaType.trim().toLowerCase() === 'application/json';
}

function ensureJsonContentType(request: FastifyRequest, reply: FastifyReply): boolean {
  if (request.method !== 'POST') return true;
  const header = request.headers['content-type'];
  if (isJsonContentType(Array.isArray(header) ? header[0] : header)) return true;
  sendError(reply, 415, 'VALIDATION_ERROR', 'application/json body required.', {
    details: { reason: 'UNSUPPORTED_MEDIA_TYPE' },
  });
  return false;
}

function ensureAuthToken(request: FastifyRequest, reply: FastifyReply): boolean {
  const expected = bridgeConfig.auth.token;
  if (!expected) return true;
  const value = request.headers[lowerCaseHeaders.auth];
  const provided = Array.isArray(value) ? value[0] : value;
  if (!provided) {
    sendError(reply, 401, 'POLICY_DENIED', 'Bridge token required to run tools.', {
      details: { reason: 'MISSING_TOKEN', header: bridgeConfig.auth.header },
    });
    return false;
  }
  if (provided !== expected) {
    sendError(reply, 401, 'POLICY_DENIED', 'Bridge token rejected.', {
      details: { reason: 'INVALID_TOKEN' },
    });
    return false;
  }
  return true;
}

function hasApprovalToken(request: FastifyRequest): boolean {
  const value = request.headers[lowerCaseHeaders.approval];
  if (Array.isArray(value)) return value.some((item) => typeof item === 'string' && item.trim().length > 0);
  return typeof value === 'string' && value.trim().length > 0;
}

async function main() {
  const fastify = Fastify({ logger: false });

  await fastify.register(rateLimit, {
    global: false,
    addHeaders: {
      'x-ratelimit-limit': true,
      'x-ratelimit-remaining': true,
      'x-ratelimit-reset': true,
    },
    errorResponseBuilder: (_request, context) =>
      buildErrorPayload('RATE_LIMITED', 'Too many requests - please slow down.', {
        details: {
          limit: context.max,
          resetInSeconds: Math.ceil(context.ttl / 1000),
        },
      }),
  });

  fastify.addHook('onRoute', (routeOptions) => {
    if (routeOptions.url?.startsWith('/artifacts')) {
      routeOptions.config = {
        ...(routeOptions.config ?? {}),
        rateLimit: bridgeConfig.rateLimit.artifacts,
      };
    }
  });

  await fastify.register(cors, {
    origin: bridgeConfig.cors.origin,
    methods: [...bridgeConfig.cors.methods],
    allowedHeaders: [...bridgeConfig.cors.allowHeaders],
    exposedHeaders: [...bridgeConfig.cors.exposeHeaders],
    credentials: bridgeConfig.cors.credentials,
    maxAge: bridgeConfig.cors.maxAgeSeconds,
  });

  const bridgePort = Number(process.env.MCP_BRIDGE_PORT || 4466);
  // Resolve MCP server cwd relative to this file location (works in dev and dist)
  const resolvedServerDir = fileURLToPath(new URL('../../mcp-server/', import.meta.url));
  const mcpServerCwd = resolvedServerDir;
  const client = new McpClient(mcpServerCwd);
  const artifactsRoot = path.join(mcpServerCwd, 'artifacts');

  await fastify.register(fastifyStatic, {
    root: artifactsRoot,
    prefix: '/artifacts/',
    index: false,
    list: false,
    decorateReply: false,
  });

  await registerArtifactEndpoints(fastify, artifactsRoot, {
    list: bridgeConfig.rateLimit.artifacts,
    detail: bridgeConfig.rateLimit.artifacts,
    files: bridgeConfig.rateLimit.artifacts,
    open: bridgeConfig.rateLimit.artifacts,
  });

  fastify.get('/health', async () => ({ status: 'ok', bridge: 'ready' }));

  fastify.get('/tools', { config: { rateLimit: bridgeConfig.rateLimit.tools } }, async () => ({
    tools: Array.from(ALLOWED_TOOLS),
  }));

  fastify.post<{ Body: RunRequestBody }>('/run', { config: { rateLimit: bridgeConfig.rateLimit.run } }, async (request, reply) => {
    if (!ensureJsonContentType(request, reply)) return;
    if (!ensureAuthToken(request, reply)) return;

    const body = request.body ?? {};
    const tool = body.tool;
    if (!tool || typeof tool !== 'string') {
      sendError(reply, 400, 'VALIDATION_ERROR', 'Tool name is required.', {
        details: { reason: 'MISSING_TOOL' },
      });
      return;
    }
    if (!ALLOWED_TOOLS.has(tool)) {
      sendError(reply, 403, 'POLICY_DENIED', `Tool not allowed: ${tool}`, {
        details: { reason: 'FORBIDDEN_TOOL', tool },
      });
      return;
    }

    const input = { ...(body.input ?? {}) };
    const applyRequested = input.apply === true;

    if (applyRequested && !WRITE_GATED_TOOLS.has(tool)) {
      sendError(reply, 403, 'POLICY_DENIED', `Tool ${tool} cannot apply changes via the bridge.`, {
        details: { reason: 'READ_ONLY_TOOL', tool },
      });
      return;
    }

    const approvalGranted = applyRequested && hasApprovalToken(request);
    if (applyRequested && WRITE_GATED_TOOLS.has(tool) && !approvalGranted) {
      sendError(reply, 403, 'POLICY_DENIED', 'Approval token required to apply changes.', {
        details: { reason: 'READ_ONLY_ENFORCED', tool },
      });
      return;
    }

    if (!approvalGranted) {
      input.apply = false;
    }

    try {
      const result = await client.run(tool, input);
      const normalized = {
        ok: true as const,
        tool,
        artifacts: result?.artifacts ?? [],
        transcriptPath: result?.transcriptPath ?? null,
        bundleIndexPath: result?.bundleIndexPath ?? null,
        diagnosticsPath: result?.diagnosticsPath ?? null,
        preview: result?.preview ?? null,
        artifactsDetail: result?.artifactsDetail ?? null,
      };
      return normalized;
    } catch (err: any) {
      const normalizedCode = normalizeRunErrorCode(err?.code);
      const inferredStatus =
        typeof err?.status === 'number'
          ? err.status
          : typeof err?.statusCode === 'number'
          ? err.statusCode
          : undefined;
      const status = inferredStatus ?? statusForCode(normalizedCode);
      const message = typeof err?.message === 'string' && err.message.length ? err.message : 'Failed to run tool.';
      const details = err?.details ?? err?.messages;
      return sendError(reply, status, normalizedCode, message, {
        details,
        incidentId: typeof err?.incidentId === 'string' ? err.incidentId : undefined,
      });
    }
  });

  async function listenWithFallback() {
    try {
      await fastify.listen({ port: bridgePort, host: '127.0.0.1' });
    } catch (err: any) {
      if (err?.code === 'EADDRINUSE' && !process.env.MCP_BRIDGE_PORT) {
        // If default port is busy and no explicit port set, pick ephemeral
        await fastify.listen({ port: 0, host: '127.0.0.1' });
      } else {
        throw err;
      }
    }
    const addr = fastify.server.address();
    const actualPort = typeof addr === 'object' && addr ? (addr as any).port : bridgePort;
    // eslint-disable-next-line no-console
    console.log(`[mcp-bridge] listening on :${actualPort}`);
  }

  await listenWithFallback();
}

main().catch((err) => {
  // eslint-disable-next-line no-console
  console.error('[mcp-bridge] fatal', err);
  process.exit(1);
});
