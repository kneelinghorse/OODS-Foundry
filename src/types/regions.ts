import type { ReactNode } from 'react';

/**
 * Canonical region contract for view composition.
 * These IDs represent the normalized slots that view extensions target.
 */

export const REGION_ORDER = [
  'globalNavigation',
  'pageHeader',
  'breadcrumbs',
  'viewToolbar',
  'main',
  'contextPanel',
] as const;

export type CanonicalRegionID = typeof REGION_ORDER[number];
export type CanonicalRegionId = CanonicalRegionID;

export type RegionMap = Record<CanonicalRegionID, ReactNode>;

const CANONICAL_REGION_SET = new Set<CanonicalRegionID>(REGION_ORDER);

export function isCanonicalRegionId(value: string): value is CanonicalRegionID {
  if (typeof value !== 'string') {
    return false;
  }

  return CANONICAL_REGION_SET.has(value.trim() as CanonicalRegionID);
}

const LEGACY_REGION_ALIASES: Readonly<Record<string, CanonicalRegionID>> = Object.freeze({
  globalnavigation: 'globalNavigation',
  nav: 'globalNavigation',
  shell: 'globalNavigation',
  header: 'pageHeader',
  pageheader: 'pageHeader',
  badges: 'pageHeader',
  meta: 'pageHeader',
  title: 'pageHeader',
  breadcrumbs: 'breadcrumbs',
  trail: 'breadcrumbs',
  actions: 'viewToolbar',
  viewtoolbar: 'viewToolbar',
  toolbar: 'viewToolbar',
  filters: 'viewToolbar',
  main: 'main',
  body: 'main',
  content: 'main',
  statusbanner: 'main',
  timeline: 'contextPanel',
  feed: 'main',
  aside: 'contextPanel',
  sidebar: 'contextPanel',
  contextpanel: 'contextPanel',
  attachments: 'contextPanel',
  comments: 'contextPanel',
  details: 'contextPanel',
  footer: 'main',
});

export function resolveCanonicalRegionId(value: string): CanonicalRegionID | null {
  if (!value) {
    return null;
  }

  const trimmed = value.trim();

  if (isCanonicalRegionId(trimmed)) {
    return trimmed;
  }

  const legacyKey = trimmed.toLowerCase();
  return LEGACY_REGION_ALIASES[legacyKey] ?? null;
}

export const LEGACY_REGION_MAP: Readonly<Record<string, CanonicalRegionID>> =
  LEGACY_REGION_ALIASES;
