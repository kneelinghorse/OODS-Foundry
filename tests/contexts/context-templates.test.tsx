import { describe, expect, it } from 'vitest';
import { renderToStaticMarkup } from 'react-dom/server';
import {
  CardView,
  DetailView,
  FormView,
  InlineView,
  ListView,
  TimelineView,
} from '../../src/contexts/index.js';
import { REGION_ORDER, type RegionMap } from '../../src/types/regions.js';

function createRegionMap(overrides: Partial<RegionMap> = {}): RegionMap {
  const base = Object.fromEntries(
    REGION_ORDER.map((region) => [region, null])
  ) as RegionMap;

  return {
    ...base,
    ...overrides,
  };
}

describe('Context templates', () => {
  describe('ListView', () => {
    it('renders required slots when regions are empty', () => {
      const regions = createRegionMap();
      const render = () => renderToStaticMarkup(<ListView regions={regions} />);

      expect(render).not.toThrow();
      expect(render()).toMatchSnapshot();
    });

    it('renders optional slots in deterministic order', () => {
      const regions = createRegionMap({
        globalNavigation: <div>Shell</div>,
        pageHeader: <header>List Header</header>,
        breadcrumbs: <nav>Trail</nav>,
        viewToolbar: <div>Filters</div>,
        main: <section>Main Area</section>,
        contextPanel: <aside>KPIs</aside>,
      });

      const markup = renderToStaticMarkup(<ListView regions={regions} />);
      expect(markup).toMatchSnapshot();
    });
  });

  describe('DetailView', () => {
    it('renders required slots when regions are empty', () => {
      const regions = createRegionMap();
      const render = () => renderToStaticMarkup(<DetailView regions={regions} />);

      expect(render).not.toThrow();
      expect(render()).toMatchSnapshot();
    });

    it('renders populated slots alongside context panel', () => {
      const regions = createRegionMap({
        globalNavigation: <div>Shell</div>,
        pageHeader: <header>Detail Header</header>,
        breadcrumbs: <nav>Trail</nav>,
        viewToolbar: <div>Actions</div>,
        main: <section>Details</section>,
        contextPanel: <aside>Timeline</aside>,
      });

      const markup = renderToStaticMarkup(<DetailView regions={regions} />);
      expect(markup).toMatchSnapshot();
    });
  });

  describe('FormView', () => {
    it('renders required slots when regions are empty', () => {
      const regions = createRegionMap();
      const render = () => renderToStaticMarkup(<FormView regions={regions} />);

      expect(render).not.toThrow();
      expect(render()).toMatchSnapshot();
    });

    it('renders optional toolbar and supporting panel', () => {
      const regions = createRegionMap({
        globalNavigation: <div>Shell</div>,
        pageHeader: <header>Form Header</header>,
        breadcrumbs: <nav>Trail</nav>,
        viewToolbar: <div>Save Actions</div>,
        main: <section>Form</section>,
        contextPanel: <aside>Helper</aside>,
      });

      const markup = renderToStaticMarkup(<FormView regions={regions} />);
      expect(markup).toMatchSnapshot();
    });
  });

  describe('TimelineView', () => {
    it('renders required slots when regions are empty', () => {
      const regions = createRegionMap();
      const render = () => renderToStaticMarkup(<TimelineView regions={regions} />);

      expect(render).not.toThrow();
      expect(render()).toMatchSnapshot();
    });

    it('renders meta inside main content with supporting panel', () => {
      const regions = createRegionMap({
        globalNavigation: <div>Shell</div>,
        breadcrumbs: <nav>Trail</nav>,
        pageHeader: <header>Timeline Header</header>,
        viewToolbar: <div>Feed Filters</div>,
        main: <section>Activity Feed</section>,
        contextPanel: <aside>Related Threads</aside>,
      });

      const markup = renderToStaticMarkup(<TimelineView regions={regions} />);
      expect(markup).toMatchSnapshot();
    });
  });

  describe('CardView', () => {
    it('renders required slots when regions are empty', () => {
      const regions = createRegionMap();
      const render = () => renderToStaticMarkup(<CardView regions={regions} />);

      expect(render).not.toThrow();
      expect(render()).toMatchSnapshot();
    });

    it('renders header and supporting content when provided', () => {
      const regions = createRegionMap({
        pageHeader: <header>Card Header</header>,
        main: <section>Card Body</section>,
        contextPanel: <aside>Annotations</aside>,
      });

      const markup = renderToStaticMarkup(<CardView regions={regions} />);
      expect(markup).toMatchSnapshot();
    });
  });

  describe('InlineView', () => {
    it('renders required slots when regions are empty', () => {
      const regions = createRegionMap();
      const render = () => renderToStaticMarkup(<InlineView regions={regions} />);

      expect(render).not.toThrow();
      expect(render()).toMatchSnapshot();
    });

    it('renders header row with badges when supplied', () => {
      const regions = createRegionMap({
        pageHeader: <header>Inline Header</header>,
        viewToolbar: <div>Badges</div>,
        main: <section>Inline Content</section>,
      });

      const markup = renderToStaticMarkup(<InlineView regions={regions} />);
      expect(markup).toMatchSnapshot();
    });
  });
});
