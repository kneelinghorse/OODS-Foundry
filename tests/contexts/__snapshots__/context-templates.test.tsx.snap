// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Context templates > CardView > renders header and supporting content when provided 1`] = `"<div class="rounded-xl border border-slate-200 bg-white p-6 shadow-sm transition-shadow dark:border-slate-800 dark:bg-slate-950 shadow-lg shadow-slate-200/50 dark:shadow-none flex flex-col gap-4" data-view="card" data-view-profile="card" data-view-has-globalnavigation="false" data-view-has-pageheader="true" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="false" data-view-has-main="true" data-view-has-contextpanel="true"><header data-region="pageHeader"><header>Card Header</header></header><main data-region="main" data-region-required="true"><section>Card Body</section></main><aside data-region="contextPanel" aria-label="Supporting context"><aside>Annotations</aside></aside></div>"`;

exports[`Context templates > CardView > renders required slots when regions are empty 1`] = `"<div class="rounded-xl border border-slate-200 bg-white p-6 shadow-sm transition-shadow dark:border-slate-800 dark:bg-slate-950 shadow-lg shadow-slate-200/50 dark:shadow-none flex flex-col gap-4" data-view="card" data-view-profile="card" data-view-has-globalnavigation="false" data-view-has-pageheader="false" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="false" data-view-has-main="false" data-view-has-contextpanel="false"><main data-region="main" data-region-required="true"></main></div>"`;

exports[`Context templates > DetailView > renders populated slots alongside context panel 1`] = `"<div data-view="detail" data-view-profile="detail" data-view-has-globalnavigation="true" data-view-has-pageheader="true" data-view-has-breadcrumbs="true" data-view-has-viewtoolbar="true" data-view-has-main="true" data-view-has-contextpanel="true"><nav data-region="globalNavigation" aria-label="Global navigation"><div>Shell</div></nav><div data-region-group="detail-body"><nav data-region="breadcrumbs" aria-label="Breadcrumbs"><nav>Trail</nav></nav><header data-region="pageHeader" data-region-required="true"><header>Detail Header</header></header><div data-region="viewToolbar" role="toolbar"><div>Actions</div></div><div data-region-group="detail-content"><main data-region="main" data-region-required="true"><section>Details</section></main><aside data-region="contextPanel" aria-label="Context panel"><aside>Timeline</aside></aside></div></div></div>"`;

exports[`Context templates > DetailView > renders required slots when regions are empty 1`] = `"<div data-view="detail" data-view-profile="detail" data-view-has-globalnavigation="false" data-view-has-pageheader="false" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="false" data-view-has-main="false" data-view-has-contextpanel="false"><div data-region-group="detail-body"><header data-region="pageHeader" data-region-required="true"></header><div data-region-group="detail-content"><main data-region="main" data-region-required="true"></main></div></div></div>"`;

exports[`Context templates > FormView > renders optional toolbar and supporting panel 1`] = `"<div data-view="form" data-view-profile="form" data-view-has-globalnavigation="true" data-view-has-pageheader="true" data-view-has-breadcrumbs="true" data-view-has-viewtoolbar="true" data-view-has-main="true" data-view-has-contextpanel="true"><nav data-region="globalNavigation" aria-label="Global navigation"><div>Shell</div></nav><div data-region-group="form-body"><nav data-region="breadcrumbs" aria-label="Breadcrumbs"><nav>Trail</nav></nav><header data-region="pageHeader" data-region-required="true"><header>Form Header</header></header><div data-region="viewToolbar" role="toolbar"><div>Save Actions</div></div><div data-region-group="form-content"><main data-region="main" data-region-required="true"><section>Form</section></main><aside data-region="contextPanel" aria-label="Context panel"><aside>Helper</aside></aside></div></div></div>"`;

exports[`Context templates > FormView > renders required slots when regions are empty 1`] = `"<div data-view="form" data-view-profile="form" data-view-has-globalnavigation="false" data-view-has-pageheader="false" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="false" data-view-has-main="false" data-view-has-contextpanel="false"><div data-region-group="form-body"><header data-region="pageHeader" data-region-required="true"></header><div data-region-group="form-content"><main data-region="main" data-region-required="true"></main></div></div></div>"`;

exports[`Context templates > InlineView > renders header row with badges when supplied 1`] = `"<div data-view="inline" data-view-profile="inline" data-view-has-globalnavigation="false" data-view-has-pageheader="true" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="true" data-view-has-main="true" data-view-has-contextpanel="false"><div data-region-group="inline-header-row"><header data-region="pageHeader"><header>Inline Header</header></header><div data-region="viewToolbar" role="toolbar"><div>Badges</div></div></div><main data-region="main" data-region-required="true"><section>Inline Content</section></main></div>"`;

exports[`Context templates > InlineView > renders required slots when regions are empty 1`] = `"<div data-view="inline" data-view-profile="inline" data-view-has-globalnavigation="false" data-view-has-pageheader="false" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="false" data-view-has-main="false" data-view-has-contextpanel="false"><main data-region="main" data-region-required="true"></main></div>"`;

exports[`Context templates > ListView > renders optional slots in deterministic order 1`] = `"<div data-view="list" data-view-profile="list" data-view-has-globalnavigation="true" data-view-has-pageheader="true" data-view-has-breadcrumbs="true" data-view-has-viewtoolbar="true" data-view-has-main="true" data-view-has-contextpanel="true"><nav data-region="globalNavigation" aria-label="Global navigation"><div>Shell</div></nav><div data-region-group="list-body"><div data-region-group="list-chrome"><nav data-region="breadcrumbs" aria-label="Breadcrumbs"><nav>Trail</nav></nav><header data-region="pageHeader" data-region-required="true"><header>List Header</header></header><div data-region="viewToolbar" data-region-required="true" role="toolbar"><div>Filters</div></div></div><div data-region-group="list-content"><main data-region="main" data-region-required="true"><section>Main Area</section></main><aside data-region="contextPanel" aria-label="Context panel"><aside>KPIs</aside></aside></div></div></div>"`;

exports[`Context templates > ListView > renders required slots when regions are empty 1`] = `"<div data-view="list" data-view-profile="list" data-view-has-globalnavigation="false" data-view-has-pageheader="false" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="false" data-view-has-main="false" data-view-has-contextpanel="false"><div data-region-group="list-body"><div data-region-group="list-chrome"><header data-region="pageHeader" data-region-required="true"></header><div data-region="viewToolbar" data-region-required="true" role="toolbar"></div></div><div data-region-group="list-content"><main data-region="main" data-region-required="true"></main></div></div></div>"`;

exports[`Context templates > TimelineView > renders meta inside main content with supporting panel 1`] = `"<div data-view="timeline" data-view-profile="timeline" data-view-has-globalnavigation="true" data-view-has-pageheader="true" data-view-has-breadcrumbs="true" data-view-has-viewtoolbar="true" data-view-has-main="true" data-view-has-contextpanel="true"><nav data-region="globalNavigation" aria-label="Global navigation"><div>Shell</div></nav><nav data-region="breadcrumbs" aria-label="Breadcrumbs"><nav>Trail</nav></nav><div data-region-group="timeline-body"><main data-region="main" data-region-required="true"><div data-region-group="timeline-meta"><header data-region="pageHeader" data-region-required="true"><header>Timeline Header</header></header><div data-region="viewToolbar" data-region-required="true" role="toolbar"><div>Feed Filters</div></div></div><div data-region-group="timeline-stream"><section>Activity Feed</section></div></main><aside data-region="contextPanel" aria-label="Context panel"><aside>Related Threads</aside></aside></div></div>"`;

exports[`Context templates > TimelineView > renders required slots when regions are empty 1`] = `"<div data-view="timeline" data-view-profile="timeline" data-view-has-globalnavigation="false" data-view-has-pageheader="false" data-view-has-breadcrumbs="false" data-view-has-viewtoolbar="false" data-view-has-main="false" data-view-has-contextpanel="false"><div data-region-group="timeline-body"><main data-region="main" data-region-required="true"><div data-region-group="timeline-meta"><header data-region="pageHeader" data-region-required="true"></header><div data-region="viewToolbar" data-region-required="true" role="toolbar"></div></div><div data-region-group="timeline-stream"></div></main></div></div>"`;
